### PR TITLE
Migrate conv2d from TH to ATen (CPU)

### PR DESCRIPTION
**Pull request by andreaskoepf**
This is a port of the SpatialConvolutionMM TH (CPU) implementation to ATen as `slow_conv2d`. In practice it is invoked  for ungrouped, non-dilated, non-float32 convolutions (e.g. float64, long, bfloat16).

- [x] unfolded_copy & unfolded_acc
- [x] forward
- [x] backward
- [x] basic sanity cross check with 1.3 impl
- [x] systematic testing
- [x] performance comparison & optimization

File used for performance testing: [benchmark_conv2d.py](https://gist.github.com/andreaskoepf/c2777b2e5e9d11610f9fc74372930527)

### CODE REVIEW

**Comment by andreaskoepf** [comment]
@ezyang Basic forward & backward functionality of conv2d should be implemented now. 

I have a few questions:

1.) I have spent two hours trying to understand an error that occurs when `addmm_()` is called from inside of an `at::parallel_for()` lambda. If I use a classic for-loop the error does not occur. I have commented the problem here: https://github.com/pytorch/pytorch/blob/c002438df4b7895bbf9eaf8baefdaa8619dbeb72/aten/src/ATen/native/ConvolutionMM2d.cpp#L281-L286
Do you have an idea what the problem might be there? Is this a known issue?

2.) In the old TH code cases for 3D inputs were handled. Can we assume that calls to conv2d only come from [_convolution_nogroup()](https://github.com/pytorch/pytorch/blob/dfe7b25eafd88130306991299574145f3333106e/aten/src/ATen/native/Convolution.cpp#L724) in Convolution.cpp? Should I remove the 3D cases? e.g.
https://github.com/pytorch/pytorch/blob/c002438df4b7895bbf9eaf8baefdaa8619dbeb72/aten/src/ATen/native/ConvolutionMM2d.cpp#L463-L466

3.) Currently of TH related names used inside LegacyNNDefinitions.cpp e.g. `at::thnn_conv2d` [here](https://github.com/pytorch/pytorch/blob/c002438df4b7895bbf9eaf8baefdaa8619dbeb72/aten/src/ATen/native/LegacyNNDefinitions.cpp#L43-L51) . Should I remove/rename those to completely remove any signs of TH?

General note regarding testing etc.: When building with NNPACK (which is enabled by default) the conv2d impl of this PR will only be used for float64 cases.

**Comment by ezyang** [comment]
> Do you have an idea what the problem might be there? Is this a known issue?

Yes, it's this one #28370 cc @albanD

Your attempted use of NoGradGuard was a good try, but I think you needed a subtly different guard here, `AutoNonVariableTypeMode`

**Comment by andreaskoepf** [comment]
### Performance test results:

Both reference master and this feature branch built with: `BLAS=MKL USE_NNPACK=0 DEBUG=0 USE_CUDA=0 BUILD_TEST=0 BUILD_CAFFE2_OPS=0 python setup.py develop`

#### Forward
The percentage value in parentheses is `new_duration / old_duration`.
The durations are averaged over 100 runs in a loop.
```
in: 3x128x256; out_ch: 32; k: 7x7; pad: 4; st: 1; new:   27.660ms old:   26.806ms (103.19%)
in: 3x128x256; out_ch: 32; k: 7x7; pad: 4; st: 2; new:    7.068ms old:    5.284ms (133.76%)
in: 3x128x256; out_ch: 32; k: 7x7; pad: 2; st: 1; new:   25.190ms old:   29.170ms (86.36%)
in: 3x128x256; out_ch: 32; k: 7x7; pad: 2; st: 2; new:    4.975ms old:    5.746ms (86.58%)
in: 3x128x256; out_ch: 32; k: 7x7; pad: 1; st: 1; new:   24.456ms old:   28.691ms (85.24%)
in: 3x128x256; out_ch: 32; k: 7x7; pad: 1; st: 2; new:    4.878ms old:    6.566ms (74.29%)
in: 3x128x256; out_ch: 32; k: 4x5; pad: 4; st: 1; new:    9.974ms old:    9.350ms (106.67%)
in: 3x128x256; out_ch: 32; k: 4x5; pad: 4; st: 2; new:    2.444ms old:    2.888ms (84.63%)
in: 3x128x256; out_ch: 32; k: 4x5; pad: 2; st: 1; new:    7.350ms old:   10.581ms (69.46%)
in: 3x128x256; out_ch: 32; k: 4x5; pad: 2; st: 2; new:    2.356ms old:    2.661ms (88.54%)
in: 3x128x256; out_ch: 32; k: 4x5; pad: 1; st: 1; new:    7.351ms old:    9.108ms (80.71%)
in: 3x128x256; out_ch: 32; k: 4x5; pad: 1; st: 2; new:    2.381ms old:    2.727ms (87.31%)
in: 3x128x256; out_ch: 32; k: 3x3; pad: 4; st: 1; new:    4.494ms old:    4.755ms (94.51%)
in: 3x128x256; out_ch: 32; k: 3x3; pad: 4; st: 2; new:    1.277ms old:    1.487ms (85.88%)
in: 3x128x256; out_ch: 32; k: 3x3; pad: 2; st: 1; new:    4.074ms old:    4.339ms (93.89%)
in: 3x128x256; out_ch: 32; k: 3x3; pad: 2; st: 2; new:    1.229ms old:    1.491ms (82.43%)
in: 3x128x256; out_ch: 32; k: 3x3; pad: 1; st: 1; new:    4.816ms old:    5.146ms (93.59%)
in: 3x128x256; out_ch: 32; k: 3x3; pad: 1; st: 2; new:    1.456ms old:    1.557ms (93.51%)
in: 3x128x256; out_ch: 64; k: 7x7; pad: 4; st: 1; new:   37.570ms old:   45.454ms (82.65%)
in: 3x128x256; out_ch: 64; k: 7x7; pad: 4; st: 2; new:    7.004ms old:    8.052ms (86.98%)
in: 3x128x256; out_ch: 64; k: 7x7; pad: 2; st: 1; new:   31.812ms old:   40.789ms (77.99%)
in: 3x128x256; out_ch: 64; k: 7x7; pad: 2; st: 2; new:    6.662ms old:    7.946ms (83.84%)
in: 3x128x256; out_ch: 64; k: 7x7; pad: 1; st: 1; new:   31.966ms old:   38.541ms (82.94%)
in: 3x128x256; out_ch: 64; k: 7x7; pad: 1; st: 2; new:    6.539ms old:    7.447ms (87.81%)
in: 3x128x256; out_ch: 64; k: 4x5; pad: 4; st: 1; new:   16.190ms old:   18.335ms (88.30%)
in: 3x128x256; out_ch: 64; k: 4x5; pad: 4; st: 2; new:    3.422ms old:    3.628ms (94.32%)
in: 3x128x256; out_ch: 64; k: 4x5; pad: 2; st: 1; new:   16.296ms old:   18.680ms (87.24%)
in: 3x128x256; out_ch: 64; k: 4x5; pad: 2; st: 2; new:    3.438ms old:    3.623ms (94.89%)
in: 3x128x256; out_ch: 64; k: 4x5; pad: 1; st: 1; new:   10.584ms old:   14.892ms (71.07%)
in: 3x128x256; out_ch: 64; k: 4x5; pad: 1; st: 2; new:    3.481ms old:    3.847ms (90.49%)
in: 3x128x256; out_ch: 64; k: 3x3; pad: 4; st: 1; new:   12.305ms old:   14.240ms (86.41%)
in: 3x128x256; out_ch: 64; k: 3x3; pad: 4; st: 2; new:    2.182ms old:    2.275ms (95.91%)
in: 3x128x256; out_ch: 64; k: 3x3; pad: 2; st: 1; new:   11.624ms old:   13.305ms (87.37%)
in: 3x128x256; out_ch: 64; k: 3x3; pad: 2; st: 2; new:    2.234ms old:    2.188ms (102.10%)
in: 3x128x256; out_ch: 64; k: 3x3; pad: 1; st: 1; new:   12.751ms old:   14.746ms (86.47%)
in: 3x128x256; out_ch: 64; k: 3x3; pad: 1; st: 2; new:    2.381ms old:    2.530ms (94.11%)
in: 3x128x256; out_ch: 24; k: 7x7; pad: 4; st: 1; new:   25.208ms old:   29.087ms (86.66%)
in: 3x128x256; out_ch: 24; k: 7x7; pad: 4; st: 2; new:    4.905ms old:    5.667ms (86.55%)
in: 3x128x256; out_ch: 24; k: 7x7; pad: 2; st: 1; new:   24.042ms old:   27.888ms (86.21%)
in: 3x128x256; out_ch: 24; k: 7x7; pad: 2; st: 2; new:    4.628ms old:    5.419ms (85.40%)
in: 3x128x256; out_ch: 24; k: 7x7; pad: 1; st: 1; new:   22.700ms old:   26.281ms (86.37%)
in: 3x128x256; out_ch: 24; k: 7x7; pad: 1; st: 2; new:    4.536ms old:    5.199ms (87.25%)
in: 3x128x256; out_ch: 24; k: 4x5; pad: 4; st: 1; new:    9.645ms old:    9.119ms (105.77%)
in: 3x128x256; out_ch: 24; k: 4x5; pad: 4; st: 2; new:    2.428ms old:    2.518ms (96.43%)
in: 3x128x256; out_ch: 24; k: 4x5; pad: 2; st: 1; new:    6.450ms old:    8.540ms (75.53%)
in: 3x128x256; out_ch: 24; k: 4x5; pad: 2; st: 2; new:    2.127ms old:    2.271ms (93.66%)
in: 3x128x256; out_ch: 24; k: 4x5; pad: 1; st: 1; new:    6.698ms old:    9.426ms (71.06%)
in: 3x128x256; out_ch: 24; k: 4x5; pad: 1; st: 2; new:    2.188ms old:    2.362ms (92.63%)
in: 3x128x256; out_ch: 24; k: 3x3; pad: 4; st: 1; new:    3.682ms old:    3.803ms (96.82%)
in: 3x128x256; out_ch: 24; k: 3x3; pad: 4; st: 2; new:    1.153ms old:    1.287ms (89.59%)
in: 3x128x256; out_ch: 24; k: 3x3; pad: 2; st: 1; new:    3.353ms old:    3.462ms (96.85%)
```

#### Forward & Backward
```
in: 3x128x256; out_ch: 32; k: 7x7; pad: 4; st: 1; new: 68.236ms  old: 90.464ms  (75.43%) 
in: 3x128x256; out_ch: 32; k: 7x7; pad: 4; st: 2; new: 16.940ms  old: 35.535ms  (47.67%)  
in: 3x128x256; out_ch: 32; k: 7x7; pad: 2; st: 1; new: 52.210ms  old: 81.809ms  (63.82%) 
in: 3x128x256; out_ch: 32; k: 7x7; pad: 2; st: 2; new: 13.037ms  old: 28.141ms  (46.33%) 
in: 3x128x256; out_ch: 32; k: 7x7; pad: 1; st: 1; new: 43.175ms  old: 77.597ms  (55.64%) 
in: 3x128x256; out_ch: 32; k: 7x7; pad: 1; st: 2; new: 15.209ms  old: 25.293ms  (60.13%) 
in: 3x128x256; out_ch: 32; k: 4x5; pad: 4; st: 1; new: 25.248ms  old: 51.951ms  (48.60%) 
in: 3x128x256; out_ch: 32; k: 4x5; pad: 4; st: 2; new: 6.600ms   old: 17.422ms  (37.88%) 
in: 3x128x256; out_ch: 32; k: 4x5; pad: 2; st: 1; new: 23.941ms  old: 48.970ms  (48.89%) 
in: 3x128x256; out_ch: 32; k: 4x5; pad: 2; st: 2; new: 7.282ms   old: 17.046ms  (42.72%) 
in: 3x128x256; out_ch: 32; k: 4x5; pad: 1; st: 1; new: 20.960ms  old: 47.474ms  (44.15%) 
in: 3x128x256; out_ch: 32; k: 4x5; pad: 1; st: 2; new: 6.786ms   old: 17.116ms  (39.65%) 
in: 3x128x256; out_ch: 32; k: 3x3; pad: 4; st: 1; new: 15.557ms  old: 40.425ms  (38.48%) 
in: 3x128x256; out_ch: 32; k: 3x3; pad: 4; st: 2; new: 4.448ms   old: 13.053ms  (34.08%) 
in: 3x128x256; out_ch: 32; k: 3x3; pad: 2; st: 1; new: 14.340ms  old: 40.789ms  (35.16%) 
in: 3x128x256; out_ch: 32; k: 3x3; pad: 2; st: 2; new: 4.264ms   old: 12.846ms  (33.19%) 
in: 3x128x256; out_ch: 32; k: 3x3; pad: 1; st: 1; new: 16.582ms  old: 42.904ms  (38.65%) 
in: 3x128x256; out_ch: 32; k: 3x3; pad: 1; st: 2; new: 5.128ms   old: 12.801ms  (40.06%) 
in: 3x128x256; out_ch: 64; k: 7x7; pad: 4; st: 1; new: 90.983ms  old: 153.852ms (59.14%) 
in: 3x128x256; out_ch: 64; k: 7x7; pad: 4; st: 2; new: 21.434ms  old: 43.201ms  (49.61%) 
in: 3x128x256; out_ch: 64; k: 7x7; pad: 2; st: 1; new: 83.509ms  old: 132.400ms (63.07%) 
in: 3x128x256; out_ch: 64; k: 7x7; pad: 2; st: 2; new: 20.804ms  old: 41.288ms  (50.39%) 
in: 3x128x256; out_ch: 64; k: 7x7; pad: 1; st: 1; new: 83.178ms  old: 133.035ms (62.52%) 
in: 3x128x256; out_ch: 64; k: 7x7; pad: 1; st: 2; new: 19.909ms  old: 38.665ms  (51.49%) 
in: 3x128x256; out_ch: 64; k: 4x5; pad: 4; st: 1; new: 48.213ms  old: 108.348ms (44.50%) 
in: 3x128x256; out_ch: 64; k: 4x5; pad: 4; st: 2; new: 10.581ms  old: 28.492ms  (37.14%) 
in: 3x128x256; out_ch: 64; k: 4x5; pad: 2; st: 1; new: 49.892ms  old: 105.356ms (47.36%) 
in: 3x128x256; out_ch: 64; k: 4x5; pad: 2; st: 2; new: 10.749ms  old: 26.893ms  (39.97%) 
in: 3x128x256; out_ch: 64; k: 4x5; pad: 1; st: 1; new: 34.778ms  old: 84.252ms  (41.28%) 
in: 3x128x256; out_ch: 64; k: 4x5; pad: 1; st: 2; new: 11.851ms  old: 26.654ms  (44.46%) 
in: 3x128x256; out_ch: 64; k: 3x3; pad: 4; st: 1; new: 38.610ms  old: 92.385ms  (41.79%) 
in: 3x128x256; out_ch: 64; k: 3x3; pad: 4; st: 2; new: 8.212ms   old: 24.380ms  (33.68%) 
in: 3x128x256; out_ch: 64; k: 3x3; pad: 2; st: 1; new: 36.672ms  old: 85.575ms  (42.85%) 
in: 3x128x256; out_ch: 64; k: 3x3; pad: 2; st: 2; new: 7.453ms   old: 22.572ms  (33.02%) 
in: 3x128x256; out_ch: 64; k: 3x3; pad: 1; st: 1; new: 36.899ms  old: 84.719ms  (43.55%) 
in: 3x128x256; out_ch: 64; k: 3x3; pad: 1; st: 2; new: 7.999ms   old: 22.506ms  (35.54%) 
in: 3x128x256; out_ch: 24; k: 7x7; pad: 4; st: 1; new: 41.803ms  old: 68.683ms  (60.86%) 
in: 3x128x256; out_ch: 24; k: 7x7; pad: 4; st: 2; new: 11.575ms  old: 24.137ms  (47.96%) 
in: 3x128x256; out_ch: 24; k: 7x7; pad: 2; st: 1; new: 39.724ms  old: 67.536ms  (58.82%) 
in: 3x128x256; out_ch: 24; k: 7x7; pad: 2; st: 2; new: 11.181ms  old: 23.042ms  (48.52%) 
in: 3x128x256; out_ch: 24; k: 7x7; pad: 1; st: 1; new: 38.679ms  old: 64.953ms  (59.55%) 
in: 3x128x256; out_ch: 24; k: 7x7; pad: 1; st: 2; new: 10.952ms  old: 22.732ms  (48.18%) 
in: 3x128x256; out_ch: 24; k: 4x5; pad: 4; st: 1; new: 18.363ms  old: 40.584ms  (45.25%) 
in: 3x128x256; out_ch: 24; k: 4x5; pad: 4; st: 2; new: 6.480ms   old: 14.890ms  (43.52%) 
in: 3x128x256; out_ch: 24; k: 4x5; pad: 2; st: 1; new: 18.262ms  old: 40.086ms  (45.56%) 
in: 3x128x256; out_ch: 24; k: 4x5; pad: 2; st: 2; new: 6.070ms   old: 14.053ms  (43.19%) 
in: 3x128x256; out_ch: 24; k: 4x5; pad: 1; st: 1; new: 18.340ms  old: 38.532ms  (47.60%) 
in: 3x128x256; out_ch: 24; k: 4x5; pad: 1; st: 2; new: 5.713ms   old: 13.475ms  (42.40%) 
in: 3x128x256; out_ch: 24; k: 3x3; pad: 4; st: 1; new: 11.566ms  old: 37.431ms  (30.90%) 
in: 3x128x256; out_ch: 24; k: 3x3; pad: 4; st: 2; new: 3.484ms   old: 10.272ms  (33.92%) 
in: 3x128x256; out_ch: 24; k: 3x3; pad: 2; st: 1; new: 11.826ms  old: 32.567ms  (36.31%) 
in: 3x128x256; out_ch: 24; k: 3x3; pad: 2; st: 2; new: 4.348ms   old: 9.947ms   (43.71%) 
in: 3x128x256; out_ch: 24; k: 3x3; pad: 1; st: 1; new: 12.485ms  old: 36.288ms  (34.41%) 
in: 3x128x256; out_ch: 24; k: 3x3; pad: 1; st: 2; new: 4.468ms   old: 10.303ms  (43.37%) 
in: 3x128x128; out_ch: 32; k: 7x7; pad: 4; st: 1; new: 25.657ms  old: 42.148ms  (60.87%) 
in: 3x128x128; out_ch: 32; k: 7x7; pad: 4; st: 2; new: 6.469ms   old: 15.853ms  (40.81%) 
in: 3x128x128; out_ch: 32; k: 7x7; pad: 2; st: 1; new: 21.295ms  old: 36.075ms  (59.03%) 
in: 3x128x128; out_ch: 32; k: 7x7; pad: 2; st: 2; new: 6.221ms   old: 13.363ms  (46.55%) 
in: 3x128x128; out_ch: 32; k: 7x7; pad: 1; st: 1; new: 21.523ms  old: 35.154ms  (61.22%) 
in: 3x128x128; out_ch: 32; k: 7x7; pad: 1; st: 2; new: 6.685ms   old: 13.664ms  (48.92%) 
in: 3x128x128; out_ch: 32; k: 4x5; pad: 4; st: 1; new: 11.006ms  old: 26.767ms  (41.12%) 
in: 3x128x128; out_ch: 32; k: 4x5; pad: 4; st: 2; new: 3.942ms   old: 9.189ms   (42.90%) 
in: 3x128x128; out_ch: 32; k: 4x5; pad: 2; st: 1; new: 10.773ms  old: 26.045ms  (41.36%) 
in: 3x128x128; out_ch: 32; k: 4x5; pad: 2; st: 2; new: 3.511ms   old: 9.093ms   (38.61%) 
```

**Comment by ezyang** [comment]
> In the old TH code cases for 3D inputs were handled. Can we assume that calls to conv2d only come from _convolution_nogroup() in Convolution.cpp? Should I remove the 3D cases

I can say that direct call to these functions is not supported :) I don't know for certain if `_convolution_nogroup` enforces this invariant, but if it does you're in the clear

**Comment by ezyang** [comment]
WIP in the title is out of date, I guess?

**Comment by andreaskoepf** [comment]
I removed the special case handling for 3D inputs and prefixed the function names with `slow_`.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 9070404:
* 0/3 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=90704041f172b458a7952f2b2f634675a6183a99).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by andreaskoepf** [comment]
ref #24775, #24775

File: aten/src/ATen/native/ConvolutionMM2d.cpp [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+                  output_height * output_width});
+  output.resize_({batch_size, n_output_plane, output_height, output_width});
+
+  at::parallel_for(0, batch_size, 0, [&](int64_t start, int64_t end) {
+    NoGradGuard no_grad;
+    AutoNonVariableTypeMode non_variable_type_mode;
+    for (int64_t t = start; t < end; t++) {
+      Tensor input_t = input[t];
+      Tensor output_t = output[t];
+      Tensor finput_t = finput[t];
```
Comment (ezyang) [inline aten/src/ATen/native/ConvolutionMM2d.cpp]: @albanD Another case of Tensor operations being done inside of a loop

File: aten/src/ATen/native/ConvolutionMM2d.cpp [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+
+  const int64_t batch_size = input.size(0);
+
+  finput.resize_({batch_size,
+                  n_input_plane * kernel_height * kernel_width,
+                  output_height * output_width});
+  output.resize_({batch_size, n_output_plane, output_height, output_width});
+
+  at::parallel_for(0, batch_size, 0, [&](int64_t start, int64_t end) {
+    NoGradGuard no_grad;
```
Comment (ezyang) [inline aten/src/ATen/native/ConvolutionMM2d.cpp]: You don't actually need this one

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/ConvolutionMM2d.cpp (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/native/ConvolutionMM2d.cpp]: cc @ilia-cher

**Pull request closed without merging**